crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/meta_net/tests.rs

```diff
@@ -749,6 +767,68 @@ async fn preflight() {
     nodes.shutdown().await;
 }
 
+#[tokio::test(flavor = "multi_thread")]
+async fn preflight_user_data_mismatch() {
+    let (recv_not, recv_wait) = notify_pair();
+
+    let mut test = Test::default();
+
+    {
+        let agent_info = crate::test_util::data::mk_agent_info(1).await;
+        test.handle_get_all_local_joined_agent_infos = Arc::new(move || {
+            let agent_info = agent_info.clone();
+            Ok(futures::future::FutureExt::boxed(async move { Ok(vec![agent_info]) }).into())
+        });
+        test.handle_put_agent_info_signed = Arc::new(move |_| {
+            recv_not.notify();
+            Ok(futures::future::FutureExt::boxed(async move { Ok(()) }).into())
+        });
+    }
+
+    let ud1 = PreflightUserData {
+        bytes: vec![1, 2, 3],
+        comparator: Box::new(|_, r| {
+            (r == &[1, 2, 3])
+                .then_some(())
+                .ok_or("preflight mismatch".into())
+        }),
+    };
+    let ud2 = PreflightUserData {
+        bytes: vec![9, 8, 7, 6, 5],
+        comparator: Box::new(|_, r| {
+            (r == &[9, 8, 7, 6, 5])
+                .then_some(())
+                .ok_or("preflight mismatch".into())
+        }),
+    };
+
+    let nodes = Setup2Nodes::new_with_user_data(test, ud1, ud2).await;
+
+    let con = nodes
+        .send1
+        .get_connection(nodes.addr2.clone(), nodes.tuning_params.implicit_timeout())
+        .await
+        .unwrap();
+
+    // This should error out because preflight failed due to user data mismatch
+    if con
+        .notify(
+            &wire::Wire::failure("Hello World!".into()),
+            nodes.tuning_params.implicit_timeout(),
+        )
+        .await
+        .is_ok()
+    {
+        // ...but if it *doesn't* error, the request should at least timeout because
+        // preflight user data doesn't match
+        tokio::time::timeout(std::time::Duration::from_millis(500), recv_wait)
+            .await
+            .unwrap_err();
+    }
```

Comment:
@neonphog can you tell what might be going on here? `con.notify` only fails sometimes, not always. Adding a 2 sec delay beforehand didn't have any effect. What is actually expected?

Reply:
I feel like notify should always error... but it's possible there is a tx5 bug.

Reply:
If it should always error, then there is a bug. Want to look into that independently of this PR?